packages/tfchain_client/src/dao.ts

```diff
@@ -0,0 +1,134 @@
+            end: moment().add(timeUntilEnd, "second"),
+            hash: hash,
+            action: hex2a(proposal.args.remark),
+            description: hex2a(daoProposal.description),
+            link: hex2a(daoProposal.link),
+            ayesProgress: this.getProgress(proposalVotes.ayes, proposalVotes.nays, true),
+            nayesProgress: this.getProgress(proposalVotes.ayes, proposalVotes.nays, false),
+          });
+        } else {
+          activeProposals.push({
+            threshold: proposalVotes.threshold,
+            ayes: proposalVotes.ayes, //[{farmId: number, weight: number}]
+            nayes: proposalVotes.nays,
+            vetos: proposalVotes.vetos,
+            end: moment().add(timeUntilEnd, "second"),
+            hash: hash,
+            action: hex2a(proposal.args.remark),
+            description: hex2a(daoProposal.description),
+            link: hex2a(daoProposal.link),
+            ayesProgress: this.getProgress(proposalVotes.ayes, proposalVotes.nays, true),
+            nayesProgress: this.getProgress(proposalVotes.ayes, proposalVotes.nays, false),
+          });
+        }
+      }
+    }
+    return { active: activeProposals, inactive: inactiveProposals };
+  }
+  private getVotesWithWeights(votes: ayesAndNayesInterface[]) {
+    return votes.reduce((total: number, vote: ayesAndNayesInterface) => {
+      return total + vote.weight;
+    }, 0);
+  }
+  private getProgress(ayes: ayesAndNayesInterface[], nayes: ayesAndNayesInterface[], typeAye: boolean) {
+    const totalAyeWeight = ayes ? this.getVotesWithWeights(ayes) : 0;
+    const totalNayeWeight = nayes ? this.getVotesWithWeights(nayes) : 0;
+    const total = totalAyeWeight + totalNayeWeight;
+    if (total > 0) {
+      if (typeAye) {
+        return (totalAyeWeight / total) * 100;
+      }
+      return (totalNayeWeight / total) * 100;
+    }
+    return 0;
+  }
+  @checkConnection
+  private async getDaoProposal(hash: { toJSON: () => any }) {
+    const proposal = await this.client.api.query.dao.proposals(hash);
+    return proposal.toJSON();
+  }
+  @checkConnection
+  private async getProposal(hash: { toJSON: () => any }) {
+    try {
+      const proposal = await this.client.api.query.dao.proposalOf(hash);
+      const res = proposal.toJSON();
+      return res;
+    } catch (error) {
+      console.warn("Couldn't decode a proposal");
+    }
+  }
+  @checkConnection
+  private async getProposalVotes(hash: any) {
+    const voting = await this.client.api.query.dao.voting(hash);
+    return voting.toJSON();
+  }
+}
+class Dao extends QueryDao {
+  constructor(public client: Client) {
+    super(client);
+    this.client = client;
+  }
+  @checkConnection
+  async vote(options: DaoVoteOptions) {
+    const nonce = await this.client.api.rpc.system.accountNextIndex(options.address);
+    return this.client.api.tx.dao
+      .vote(options.farmId, options.hash, options.approve)
+      .signAndSend(this.client.keypair, { nonce });
```

Comment:
client is already creating the nonce and call `signAndSend` after. check how to implement [here](https://github.com/threefoldtech/tfgrid-sdk-ts/blob/development/packages/tfchain_client/src/contracts.ts#L252-L254) for example